MuZero algorithm

emailweixu opened the pull request:
This PR contains two commits. The first one is the full MuZero algorithm. The second one is for speeding up MCTSAlgorithm.

emailweixu committed:
MuZero algorithm

emailweixu committed:
Memory efficient frame stack.

The new FrameStacker achieved its purpose by a different transform_experience() function to directly operate on ReplayBuffer.
Also introduced DataTransformer and unify frame stack, reward_shaping_fn and observation_transformer into DataTransformer.

emailweixu committed:
Faster MCTSAlgorithm

emailweixu committed:
Option for profiling

emailweixu committed:
Address comments

emailweixu committed:
Merge branch 'pytorch' of https://github.com/HorizonRobotics/alf into PR_data_transformer

emailweixu committed:
Bug fixes

Haichao-Zhang commented on alf/algorithms/mcts_algorithm_test.py:
```diff
@@ -27,29 +27,23 @@ class TicTacToeModel(MCTSModel):
-            1, 0
-        ])
-        self._actions = torch.arange(9, dtype=torch.int32).unsqueeze(0)
+        super().__init__()
```
TicTacToeModel has some repeated code with TicTacToeEnv. Is it a good idea to reduce it, for example, by leveraging an env instance for implementing TicTacToeModel?

emailweixu replied:
I am aware of that. But decided not to do that for two reasons:
1. It will make code structure more complex
2. This is not likely to be reused anyway.

emailweixu committed:
Fix create_control_mcts

Haichao-Zhang commented on alf/algorithms/mcts_algorithm.py:
```diff
@@ -324,15 +282,47 @@ def _print_tree(self, trees: _MCTSTree, b):
+            - search_paths: [T, B] int64 matrix, where T is the max length of the
+                search paths. For paths whose length is shorter than T, search_path
+                is padded with the last node index of that path.
+            - path_lenghts: [B] vector, length of each search path
```
path_lengths

emailweixu replied:
fixed

Haichao-Zhang commented on alf/algorithms/mcts_algorithm.py:
```diff
@@ -490,3 +494,61 @@ def visit_softmax_temperature(num_moves):
+                        action_spec,
+                        num_simulations=50,
+                        debug_summaries=False):
+    """Helper function for creating MCTSAlgorithm for atari games."""
```
atari games -> control

emailweixu replied:
fixed

Haichao-Zhang commented on alf/algorithms/muzero_algorithm.py:
```diff
@@ -0,0 +1,265 @@
+            action_spec (BoundedTensorSpec): representing the actions.
+            model_ctor (Callable): will be called as
+                ``model_ctor(observation_spec=?, action_spec=?, debug_summaries=?)``
+                to constrcut the model. The model should follow the interface
```
construct

emailweixu replied:
fixed

Haichao-Zhang commented on alf/algorithms/muzero_algorithm.py:
```diff
@@ -0,0 +1,265 @@
+                the discounted return. -1 means to bootstrap to the end of the game.
+                Can only used for environments whose rewards are zero except for
+                the last step as the current implmentation only use the reward
+                at the last step to caculate the return.
```
caculate  -> calculate

emailweixu replied:
fixed

Haichao-Zhang commented on alf/algorithms/muzero_algorithm.py:
```diff
@@ -0,0 +1,265 @@
+            model_output = self._model.recurrent_inference(
+                model_output.state, info.action[:, i, ...])
+            model_output = model_output._replace(
+                state=scale_gradient(model_output.state, 0.5))
```
Might be useful to make this 0.5 to be configurable?

emailweixu replied:
added.

Haichao-Zhang commented on alf/algorithms/mcts_algorithm_test.py:
```diff
@@ -62,77 +56,104 @@ def initial_inference(self, observation):
+        board_full = (board == 0).sum(dim=(1, 2)) == 0
+        B = torch.arange(board.shape[0]).unsqueeze(-1).unsqueeze(-1)
+        lines = board[B, self._line_y, self._line_x]
+        player1_won = ((lines == -1).sum(dim=2) == 3).any(dim=1)
```
player1_won / player2_won -> player0_won / player1_won

emailweixu replied:
updated.

emailweixu committed:
Address comments

Separate value minmax stats, normalize action_prob

emailweixu committed:
Merge branch 'pytorch' of https://github.com/HorizonRobotics/alf into PR_MuZero

emailweixu committed:
Fix merge error

emailweixu committed:
Fix bug of normalize_value

emailweixu committed:
Fix update_value_stats at wrong time

Also include
1. Better handle of incomplete episode by not treating dicount!=0 as episode end
2. State dependent std and scale distribution for continuous action
3. More summary information
